src/t_zset.c

```diff
@@ -2358,46 +2380,68 @@ void zunionInterGenericCommand(client *c, robj *dstkey, int op) {
         /* We now are aware of the final size of the resulting sorted set,
          * let's resize the dictionary embedded inside the sorted set to the
          * right size, in order to save rehashing time. */
-        dictExpand(dstzset->dict,dictSize(accumulator));
+        if (dstkey) dictExpand(dstzset->dict,dictSize(accumulator));
 
         while((de = dictNext(di)) != NULL) {
             sds ele = dictGetKey(de);
             score = dictGetDoubleVal(de);
-            znode = zslInsert(dstzset->zsl,score,ele);
-            dictAdd(dstzset->dict,ele,&znode->score);
+            if (!dstkey) {
+                cardinality++;
+                addReplyBulkCBuffer(c,ele,sdslen(ele));
+                if (withscores) {
+                    cardinality++;
+                    addReplyDouble(c, score);
+                }
+                sdsfree(ele);
```

Comment:
i think we need to return the elements sorted!
maybe the easy thing to do is build a zset object (calling zslInsert), and then destruct it after responding with it's contents.
note that `sunionDiffGenericCommand` does a similar thing (calling `createIntsetObject`) and even here we already create a dict, so the difference between creating a dict and zset is not that big (coding wise).
unless we think there's an important use case of getting non-sorted results in which case the sorting complexity is unwanted (but should probably be optional). @itamarhaber @guybe7 WDYT?

IIUC, in the case of SET_OP_INTER this happens implicitly, so the current code is already good.

Reply:
@oranagra yes, the output should be sorted (it's worth the "cost")

Reply:
@oranagra @guybe7 Thank you for your review

> IIUC, in the case of SET_OP_INTER this happens implicitly, so the current code is already good.

At present, SET_OP_INTER is not ordered, because the score of the same field may change the original order after multiple zset aggregates, for example:
```
127.0.0.1:6379> zadd zset1 0 a 1 b 3 c
(integer) 3
127.0.0.1:6379> zadd zset2 4 a 1 b 0 f
(integer) 3
127.0.0.1:6379> zinter 2 zset1 zset2 withscores
1) "a"
2) "4"
3) "b"
4) "2"
127.0.0.1:6379>
```

An approach similar to SINTER/SUNION is to use zset, `createZsetObject` by default, and finally release if don't need STORE. see [here](https://github.com/redis/redis/blob/unstable/src/t_set.c#L1083).

Reply:
i think that's what we need to do. unless @itamarhaber or anyone else says that there are many valid use cases that don't need that sorting and would suffer from the performance impact. if that is the case then we should probably have an optional SORT flag.

Reply:
Thanks, I also agree with createZsetObject by default.

Reply:
already updated. please have a look when you have time.